File: src/main/java/boj/part2/needToReview/Lim1697.java
package boj.part2.needToReview;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.LinkedList;
import java.util.Queue;
import java.util.StringTokenizer;

public class Lim1697 {	
	public static void main(String[] args) throws Exception {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());
		int n = Integer.parseInt(st.nextToken());
		int k = Integer.parseInt(st.nextToken());
		
		int min = Math.abs(n - k);
		
		if(n < k) {
			min = bfs(n, k);
		}
		
		System.out.println(min);
	}

	private static int bfs(int n, int k) {
		boolean[] visited = new boolean[200001];
		visited[n] = true;
		Queue<Point> queue = new LinkedList<Point>();
		queue.offer(new Point(n, 0));
		int min = 0;
		
		while(!queue.isEmpty()) {
			Point poll = queue.poll();
			int idx = poll.getIdx();
			int time = poll.getTime();
			
			if(idx == k) {
				min = time;
				break;
			}
			
			if(idx < k) {
				if(idx != 0 && !visited[idx * 2]) {
					visited[idx * 2] = true;
					queue.offer(new Point(idx * 2, time + 1));
				}
				
				if(!visited[idx + 1]) {
					visited[idx + 1] = true;
					queue.offer(new Point(idx + 1, time + 1));
				}
			}
					
			if(idx > 0 && !visited[idx - 1]) {
				visited[idx - 1] = true;
				queue.offer(new Point(idx - 1, time + 1));
			}		
		}
		
		return min;
	}
	
	private static class Point {
		private int idx;
		private int time;
		
		public Point(int idx, int time) {
			this.idx = idx;
			this.time = time;
		}

		public int getIdx() {
			return idx;
		}

		public int getTime() {
			return time;
		}
	}
}

File: src/main/java/boj/part7/Lim1697.kt
package boj.part7

import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.*

fun main() {
    val br = BufferedReader(InputStreamReader(System.`in`))
    val (n, k) = br.readLine().split(" ").map { it.toInt() }
    val visited = BooleanArray(200_000)

    class Subin(
        val x: Int,
        val time: Int
    ) {
        fun getNexts(): Set<Subin> {
            val result = mutableSetOf<Subin>()

            if (x - 1 >= 0 && !visited[x - 1]) {
                result.add(Subin(x - 1, time + 1))
                visited[x - 1] = true
            }
            if (x + 1 <= k && !visited[x + 1]) {
                result.add(Subin(x + 1, time + 1))
                visited[x + 1] = true
            }
            if (x * 2 != 0 && x * 2 < k * 2 && !visited[x * 2]) {
                result.add(Subin(x * 2, time + 1))
                visited[x * 2] = true
            }

            return result
        }
    }

    val queue: Queue<Subin> = LinkedList()
    queue.offer(Subin(n, 0))
    visited[n] = true

    while (queue.isNotEmpty()) {
        val polled = queue.poll()
        if (polled.x == k) {
            println(polled.time)
            return
        }

        queue.addAll(polled.getNexts())
    }
}
